pydatastructs/linear_data_structures/algorithms.py

```diff
@@ -360,3 +363,246 @@ def matrix_multiply_parallel(matrix_1, matrix_2, num_threads):
+            expects a number
+    """
+    value = float('inf')
+    if maximize:
+        value = float('-inf')
+    return [[value for a in range(0, columns+1)] for a in range(0, rows+1)]
+
+
+def _optimal_grouping_rec(object_arr, cost_storage: List[List[int]], solution_matrix: List[List[int]], maximize_prob: bool, min_compare_len: int, lookup_index: List[int], get_lookup_fn, cost_fn):
+    """
+    Helper function for optimal_grouping function
+    """
+
+    # gets the present value at the present index
+    present_value = _get_value_opt_group(cost_storage, lookup_index)
+    # return the present value if it is not infinity
+    if _compare_opt_group(maximize_prob, present_value):
+        return present_value
+
+    # get the start and end indices where end index depends on the min_compare_len
+    start_index = lookup_index[0]
+    end_index = lookup_index[1]+1-(min_compare_len-1)
+
+    if start_index == end_index or start_index > end_index:
+        cost = cost_fn(object_arr, lookup_index, start_index)
+        if _compare_opt_group(maximize_prob, cost, present_value):
+            _set_value_opt_group(cost_storage, lookup_index, cost)
+            _set_value_opt_group(solution_matrix, lookup_index, start_index)
+            present_value = cost
+
+    for i in range(start_index, end_index):
+
+        # get indices for left recursion tree
+        left_rec_indices = get_lookup_fn('before', lookup_index, i)
+        _test_lookup_function(left_rec_indices, lookup_index)
+
+        cost = _optimal_grouping_rec(object_arr, cost_storage, solution_matrix, maximize_prob,
+                                    min_compare_len, left_rec_indices, get_lookup_fn, cost_fn)
+
+        # get indices for right recursion tree
+        right_rec_indices = get_lookup_fn('after', lookup_index, i)
+        _test_lookup_function(right_rec_indices, lookup_index)
+
+        cost = cost+_optimal_grouping_rec(object_arr, cost_storage, solution_matrix, maximize_prob,
+                                         min_compare_len, right_rec_indices, get_lookup_fn, cost_fn)
+
+        # get cost for present partition
+        cost = cost+cost_fn(object_arr, lookup_index, i)
+
+        # update the values if this is the best solution until now
+        if _compare_opt_group(maximize_prob, cost, present_value):
+            _set_value_opt_group(cost_storage, lookup_index, cost)
+            _set_value_opt_group(solution_matrix, lookup_index, i)
+            present_value = cost
+
+    return present_value
+
+
+def _test_lookup_function(lookup_index: List[int], input_index: List[int]):
+    if lookup_index is None:
+        raise TypeError(
+            'Check lookup_function: returning wrong type should return an array of start and end index')
+
+    if lookup_index.__len__() < 2:
+        raise ValueError(
+            'Check lookup_function:lookup index should at least have 2 integer items, first specifying the start and second specifying the last indices')
```

Comment:
The tests should be in https://github.com/codezonediitj/pydatastructs/blob/master/pydatastructs/linear_data_structures/tests/test_algorithms.py